diff --git a/surf.c b/surf.c
index 2b54e3c..153c164 100644
--- a/surf.c
+++ b/surf.c
@@ -1824,6 +1868,8 @@ clipboard(Client *c, const Arg *a)
 		gtk_clipboard_set_text(gtk_clipboard_get(
 		                       GDK_SELECTION_PRIMARY), c->targeturi
 		                       ? c->targeturi : geturi(c), -1);
+  	Arg a = {.v = (const char *[]){ "/bin/sh", "-c", "notify-send Copied: $0", c->targeturi ? c->targeturi : geturi(c) , NULL } };
+		spawn(c, &a);
 	}
 }
 
