--- a/surf.c	2019-03-21 15:54:58.740333850 -0500
+++ b/surf.c	2019-03-21 15:54:55.320333802 -0500
@@ -1119,8 +1120,12 @@
 /* For more interesting settings, have a look at
  * http://webkitgtk.org/reference/webkit2gtk/stable/WebKitSettings.html */
 
+		char *ua;
+		ua = getenv("SURF_USERAGENT");
 		if (strcmp(fulluseragent, "")) {
 			webkit_settings_set_user_agent(settings, fulluseragent);
+		} else if (strcmp(ua, "")) {
+			webkit_settings_set_user_agent(settings, ua);
 		} else if (surfuseragent) {
 			webkit_settings_set_user_agent_with_application_details(
 			    settings, "Surf", VERSION);
