diff --git a/surf.c b/surf.c
index f01a91c..5762e65 100644
--- a/surf.c
+++ b/surf.c
@@ -224,6 +224,12 @@ static void togglecookiepolicy(Client *c, const Arg *a);
 static void toggleinspector(Client *c, const Arg *a);
 static void find(Client *c, const Arg *a);
 
+char* hoveredlink = "";
+void hlclipcopy(Client *c, const Arg *a)
+{
+	gtk_clipboard_set_text(gtk_clipboard_get(GDK_SELECTION_PRIMARY), hoveredlink, -1);
+}
+
 /* Buttons */
 static void clicknavigate(Client *c, const Arg *a, WebKitHitTestResult *h);
 static void clicknewwindow(Client *c, const Arg *a, WebKitHitTestResult *h);
@@ -1262,6 +1268,7 @@ winevent(GtkWidget *w, GdkEvent *e, Client *c)
 	switch (e->type) {
 	case GDK_ENTER_NOTIFY:
 		c->overtitle = c->targeturi;
+		hoveredlink = c->targeturi;
 		updatetitle(c);
 		break;
 	case GDK_KEY_PRESS:
